File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DangerousGetHandleShouldNotBeCalled.cs
using System;
using System.Runtime.InteropServices;
using Microsoft.Win32;

namespace Tests.Diagnostics
{
    class Program
    {
        private static IntPtr GetRegistryKeyHandle(RegistryKey rKey)
        {
            Type registryKeyType = typeof(RegistryKey);

            System.Reflection.FieldInfo fieldInfo =
                registryKeyType.GetField("hkey", System.Reflection.BindingFlags.NonPublic |
                System.Reflection.BindingFlags.Instance);

            if (fieldInfo != null)
            {
                SafeHandle handle = (SafeHandle)fieldInfo.GetValue(rKey);
                IntPtr dangerousHandle = handle.DangerousGetHandle(); // Noncompliant {{Refactor the code to remove this use of 'SafeHandle.DangerousGetHandle'.}}
//                                              ^^^^^^^^^^^^^^^^^^
                IntPtr dangerousHandle2 = (((handle))).DangerousGetHandle(); // Noncompliant
                IntPtr? dangerousHandle3 = handle?.DangerousGetHandle(); // Noncompliant
                DangerousGetHandle();
                return dangerousHandle;
            }

            return IntPtr.Zero;
        }

        private static IntPtr DangerousGetHandle()
        {
            return IntPtr.Zero;
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/DangerousGetHandleShouldNotBeCalled.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Runtime.InteropServices

Module Tests

  Sub Dangerous(fieldInfo As System.Reflection.FieldInfo)
    Dim handle As SafeHandle = CType(fieldInfo.GetValue(fieldInfo), SafeHandle)
    Dim dangerousHandle As IntPtr = handle.DangerousGetHandle() ' Noncompliant {{Refactor the code to remove this use of 'SafeHandle.DangerousGetHandle'.}}
'                                          ^^^^^^^^^^^^^^^^^^
    handle.DangerousGetHandle() ' Noncompliant
  End Sub

  Public Sub Test(bar As Bar)
    Dim dangerousHandle As IntPtr = bar.DangerousHandle() ' compliant
  End Sub

  Class Bar
    Sub Test()
      DangerousHandle()
    End Sub
    Function DangerousHandle() As IntPtr
    End Function
  End Class
End Module
